File: ICS/hillcipher.cpp
#include<bits/stdc++.h>
using namespace std;
#define mod 1000000007
#define rep(i,k,n) for(int i = k;i < n;++i)
#define pb push_back
#define mk make_pair
#define f first
#define s second


string HillCipher(string message , string key){
	// Convert the keystring to matrix .
	// Convert the message string to matrix .
	// Multiply the given matrices .
	// modulo 26 to the result .
	int message_len = message.length();
	int key_len = key.length();

	if(sqrt(key_len) != message_len){
		return "";
	}


	int keyCount = 0;

	int keyMatrix[message_len][message_len];
	for(int i = 0;i < message_len;i++){
		for(int j = 0;j < message_len;j++){
			keyMatrix[i][j] = (key[keyCount++]%65);
		}
	}

	int messageCount = 0;
	int messageMatrix[message_len][1];
	for(int i = 0;i < message_len;i++){
		for(int j = 0;j < 1;j++){
			messageMatrix[i][j] = (message[messageCount++]%65);
		}
	}

	int cipherMatrix[message_len][1];

	for(int i = 0;i < message_len;i++){
		for(int j = 0;j < 1;j++){
			cipherMatrix[i][j] = 0;
			for(int x = 0;x < message_len;x++){
				cipherMatrix[i][j] += keyMatrix[i][x] * messageMatrix[x][j];
			}
			cipherMatrix[i][j] = (cipherMatrix[i][j]%26);
		}
	}

	string ciphertext;

	for(int i = 0;i < message_len;i++){
		ciphertext += cipherMatrix[i][0] + 65;
	}

	return ciphertext;

}


int main(){
	string message , key;
	cout<<"Enter the message"<<endl;
	cin>>message;
	cout<<"Enter the key";
	cin>>key;

	cout<<HillCipher(message , key);

	// cout<<"The ciphtertext is : "<<HillCipher(message , key)<<endl;	
}

File: ICS/HillCipher.java
import java.util.*;
import java.util.Arrays;
import java.util.Scanner;
import java.io.*;

class Solution{
	String hillCipher(String message , String key){
		int messageLen = message.length();
		int keyLen = key.length();

		if(Math.sqrt(keyLen) != messageLen){
			return "";
		}

		int[][] messageMatrix = new int[messageLen][1];
		int[][] keyMatrix = new int[messageLen][messageLen];
		int[][] cipherMatrix = new int[messageLen][1];

		int keyCount = 0 , messageCount = 0;

		for(int i = 0;i < messageLen;i++){
			for(int j = 0;j < messageLen;j++){
				keyMatrix[i][j] = (key.charAt(keyCount++)%65);
			}
		}

		for(int i = 0;i < messageLen;i++){
			for(int j = 0;j < 1;j++){
				messageMatrix[i][j] = (message.charAt(messageCount++)%65);
			}
		}

		for(int i = 0;i < messageLen;i++){
			cipherMatrix[i][0] = 0;
			for(int x = 0;x < messageLen;x++){
				cipherMatrix[i][0] += keyMatrix[i][x] * messageMatrix[x][0];
			}
			cipherMatrix[i][0] = (cipherMatrix[i][0]%26);
		}

		String cipherText ="";

		for(int i = 0;i < messageLen;i++){
			cipherText += (char)(cipherMatrix[i][0] + 65);
		}

		return cipherText;
	}
}

class HillCipher{
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	public static void main(String[] args) throws IOException{
		// System.out.println();
		String message , key;
		System.out.println("Enter the message");
		message = br.readLine();
		System.out.println("Enter the key");
		key = br.readLine();
		Solution obj = new Solution();
		System.out.println(obj.hillCipher(message , key));
	}
}